serde/src/java/org/apache/hadoop/hive/serde2/avro/InstanceCache.java

```diff
@@ -51,20 +51,19 @@ public Instance retrieve(SeedObject hv) throws AvroSerdeException {
    * Retrieve (or create if it doesn't exist) the correct Instance for this
    * SeedObject using 'seenSchemas' to resolve circular references
    */
-  public synchronized Instance retrieve(SeedObject hv,
-      Set<SeedObject> seenSchemas) throws AvroSerdeException {
+  public Instance retrieve(SeedObject hv, Set<SeedObject> seenSchemas)
+    throws AvroSerdeException {
     if(LOG.isDebugEnabled()) LOG.debug("Checking for hv: " + hv.toString());
 
     if(cache.containsKey(hv)) {
```

Comment:
Please leverage `Map#computeIfAbsent` here

Reply:
I had left [a comment about this on HIVE-19261](https://issues.apache.org/jira/browse/HIVE-19261?focusedCommentId=16929610&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16929610).

I think `makeInstance()` makes reentrant calls into `InstanceCache::retrieve()`. 
`computeIfAbsent()` expects that `cache` not be modified in the lambda. 
Calling calling `makeInstance()` from `computeIfAbsent()` might lead to a hang. I ran into this when testing an older version of this patch, some months ago.

It's possible that the reentrant call no longer occurs. I'd check this before using `computeIfAbsent()`.

Reply:
I poked around a bit, nothing is obviously reentrant.  Do you know where this is happening?

Reply:
> I poked around a bit, nothing is obviously reentrant.

Sorry for the delay. I remember this being convoluted... Let me see if I can recall this. Just a moment.

Reply:
Ok, I think I have it.

[`InstanceCache::retrieve()`](https://github.com/apache/hive/blob/f37c5de6c32b9395d1b34fa3c02ed06d1bfbf6eb/serde/src/java/org/apache/hadoop/hive/serde2/avro/InstanceCache.java#L66) modifies the cache here:
```java
    Instance instance = makeInstance(hv, seenSchemas);
    cache.put(hv, instance);
    return instance;
```
We would like to use `computeIfAbsent()` instead:
```java
cache.computeIfAbsent(hv, ()->makeInstance(hv, seenSchemas));
```
If `makeInstance()` ends up modifying `cache`, (e.g. by calling `retrieve()`), that would cause undefined behaviour in`computeIfAbsent()`.

[`SchemaToTypeInfo`](https://github.com/apache/hive/blob/f37c5de6c32b9395d1b34fa3c02ed06d1bfbf6eb/serde/src/java/org/apache/hadoop/hive/serde2/avro/SchemaToTypeInfo.java#L116) implements `InstanceCache::makeInstance()` as follows:
```java
  static InstanceCache<Schema, TypeInfo> typeInfoCache = new InstanceCache<Schema, TypeInfo>() {
                                  @Override
                                  protected TypeInfo makeInstance(Schema s,
                                      Set<Schema> seenSchemas)
                                      throws AvroSerdeException {
                                    return generateTypeInfoWorker(s, seenSchemas); // <---- HERE!
                                  }
                                };
```

Please note that [`SchemaToTypeInfo::generateTypeInfo()`](https://github.com/apache/hive/blob/f37c5de6c32b9395d1b34fa3c02ed06d1bfbf6eb/serde/src/java/org/apache/hadoop/hive/serde2/avro/SchemaToTypeInfo.java#L186) happens to call `InstanceCache::retrieve()`:
```java
  public static TypeInfo generateTypeInfo(Schema schema,
      Set<Schema> seenSchemas) throws AvroSerdeException {
    // ...
    return typeInfoCache.retrieve(schema, seenSchemas);
  }
```
Hence, the recursive call. :/

Reply:
OK.  That makes sense.  Thanks for running that down.

If you have a schema composed of several sub-schemas, then storing the root schema in the cache will trigger the creation of all the sub-schemas and those will also be inserted into the cache while the the initial insert still has the lock.

With that said, I don't think it's worth changing the current construct.  Unless there is some sort of demonstrable performance value here, I think it's confusing to use a `ConcurrentHashMap` but not be able to use the `computeX()` routine.   If you would like, you can clean up the logging, shrink the synchronized block (instead of the entire method), and make some simplifications like...

```
Instance instance = cache.get(hv);
if (o == null) {
    instance = makeInstance(hv, seenSchemas);
    cache.put(hv, instance);
} 
return instance;
```

In that way, the synchronized lock will be held for shorter intervals.  Oh, and obviously need JavaDoc here to explain the reentrant piece.